refactor: impl canonical network key

oetyng opened the pull request:
This was originally part of PR #212, but can be broken out to its own PR.

We've been keeping our `XorName` closeness logic in parallel with the kad closeness, i.e. mapping `PeerId` to our concept of id. We did some hashing that seems it would lead to us having one view of what is close peer, while kad had another view of it. 
Maybe it wasn't actually happening, but the two logics are there side by side, and there is not so much clarity there.

As I was looking at data replication, I got more clarity there, and I found what I believe to be the way to use only the kad closeness logic, and introducing some simplification and type safety.

This PR is an intermediate step and further clarity can be reached.

***

The closeness logic exists in `libp2p::kad::KadBucketKey`. It does all that work that we had our custom code for. We already use it for close calc, but there was not so much type safety to ensure we use only that, and a bit confusing where `XorName` is used, why and how - and how closeness of `PeerId` fits into closeness of `XorName`.

`KadBucketKey` (in this PR aliased as `Key`) works on a bit vector - which is also what `PeerId` is a wrapping type for, and it has a `preimage` which is what we pass in, and it has a key which is the `SHA256 digest` of the `preimage`.

So in this PR a canonical `NetworkKey` is implemented for cohesion. It holds the bit vector of either `PeerId` or XorName, without any modifications to it.
So it has two enum variants, `PeerId` and `XorName`. And it means we leave all bit vectors untouched before they get mapped to `KBucketKey`. 

The `NetworkKey` is now what is passed around in the paths where we use dst/distance/closeness metrics.

This should keep the distance metrics of all our items the same, and also clear up when to use what, how they relate etc.

oetyng committed:
refactor!: encaps peerid bytes instead of hashing

- This allows back and forth conversion which will be useful when
using libp2p::kad::KBucketKey closeness calcs.

oetyng committed:
feat: impl NetworkKey

- This will be useful for our needs to locate peers and data by a
common metric using the libp2p::kad::KBucketKey.

oetyng committed:
refactor!: impl the canonical network key

- This introduces a uniform and consistent usage of the KBucketKey logic
for distance calculation, whereby our own XorName calcs can and shall be
discontinued.

maqi commented on safenode/src/domain/storage/address/mod.rs:
```diff
@@ -22,23 +24,26 @@ use serde::{Deserialize, Serialize};
     Register(RegisterAddress),
     ///
     Spend(DbcAddress),
+    /// The new general identifier for a items in the network.
```
for an item in the network

However I feel this is bit tricky (as `item` covers Chunk/Register/Spend) which may cause some misuse that:
on one side we use address of `Chunk`, `Register` , `Spend`, but on other side use `Network` key.

Shall we `force` to use just one via type safe?

oetyng replied:
Yeah, I wasn't quite sure, so felt this is an intermediate step.

Thinking about it now... if good from get go or can wait..

oetyng replied:
The differentiation in DataAddress is maybe not necessary since we anyway always know what the action (query/cmd) is by the name of it..

Maybe on the benefit side is that it's a bit clearer in code. But I don't think there are many places where we could be confused about it.

oetyng replied:
I will defer a potential DataAddress update to a follow up PR.

maqi commented on safenode/src/network/event.rs:
```diff
@@ -85,7 +85,9 @@ pub enum NetworkEvent {
-    PeerAdded,
+    PeerAdded(PeerId),
+    /// Emitted when the DHT is updated
+    PeersAdded(Vec<PeerId>),
```
PeerAdded can be PeersAdded with just one entry?

oetyng replied:
Could be, absolutely. Fixing.

oetyng replied:
Fixed.

maqi commented on safenode/src/node/api.rs:
```diff
@@ -129,20 +129,28 @@ impl Node {
             }
+            NetworkEvent::PeersAdded(peers) => {
+                self.events_channel.broadcast(NodeEvent::ConnectedToNetwork);
+                for peer in peers {
```
the purpose of this get_closest block is just to get the node try to discover other nodes.
So, no need to `get_closest for each peer` , just pick one (or a buset, say number_nodes/8 ) random peer to get_closest shall be enough.

oetyng replied:
Will fix!

oetyng replied:
But how many could actually be gotten in such a vec of peers? Is it many many?

oetyng replied:
Fixed.

maqi commented on safenode/src/node/api.rs:
```diff
@@ -129,20 +129,28 @@ impl Node {
                 self.handle_request(req, channel).await?
             }
-            NetworkEvent::PeerAdded => {
+            NetworkEvent::PeerAdded(peer) => {
```
I think merge `PeerAdded` and `PeersAdded` will save a lot of traffic on `get_closest`

oetyng replied:
Fixed!

oetyng committed:
refactor: small subset of peers for close query

- This will yield 1 peer for 0-11, and 2 for 11-19. This seemed
sufficient, as libp2p discover others itself during the call to get
closest.
